File: 08/d08-1.py
import numpy as np

with open("input.dat", "r") as fp:
    heights = np.array(
        [[int(c) for c in line.strip()] for line in fp], dtype=int)

nn = heights.shape[0]
score = np.zeros((nn, nn), dtype=int)

visible = np.ones((nn, nn), dtype=bool)
for irow in range(1, score.shape[0] - 1):
    for icol in range(1, score.shape[1] - 1):
        visible[irow, icol] = (
            np.all(heights[irow, icol] > heights[irow, : icol])
            or np.all(heights[irow, icol] > heights[irow, icol + 1 :])
            or np.all(heights[irow, icol] > heights[: irow, icol])
            or np.all(heights[irow, icol] > heights[irow + 1 :, icol])
        )

print(f"Nr. of visible trees: {np.count_nonzero(visible)}")

File: 08/d08-1.f90
program d08_1
  implicit none

  integer, parameter :: max_line_len = 1024

  character(max_line_len) :: line
  integer :: fp, nn, ii, jj
  integer, allocatable :: heights(:,:)
  logical, allocatable :: visible(:,:)

  open(newunit=fp, file="input.dat", action="read")
  read(fp, "(a)") line
  nn = len_trim(line)
  rewind(fp)
  allocate(heights(nn, nn))
  do ii = 1, nn
    read(fp, "(*(i1))") heights(:, ii)
  end do
  close(fp)

  allocate(visible(nn, nn), source=.true.)
  do ii = 2, nn - 1
    do jj = 2, nn - 1
      visible(jj, ii) = &
          & all(heights(jj, ii) > heights(jj, : ii - 1))&
          & .or. all(heights(jj, ii) > heights(jj, ii + 1 :))&
          & .or. all(heights(jj, ii) > heights(: jj - 1, ii))&
          & .or. all(heights(jj, ii) > heights(jj + 1 :, ii))
    end do
  end do

  print "(a, i0)", "Nr. of visible trees: ", count(visible)

end program d08_1
